testing/ase/test_api_c_ase.cpp

```diff
@@ -0,0 +1,214 @@
+*
+*/
+TEST(sim_sw_ase, ase_common_03) {
+	struct _fpga_event_handle ehandle;
+
+	ehandle.magic = 0xFFFFFFFF;
+	ehandle.lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
+	EXPECT_EQ(FPGA_INVALID_PARAM, event_handle_check_and_lock(&ehandle));
+}
+
+/**
+ * @test       fpga_version
+ *
+ * @brief      When the parameter fpga_version* to fpgaGetOPAECVersion() is nullptr, 
+ *             the function returns FPGA_INVALID_PARAM.
+ *             If the parameter fpga_version* to fpgaGetOPAECVersion() is valid,
+ *             the function returns FPGA_OK.
+ *             When the parameter version_str* to fpgaGetOPAECVersionString() is nullptr, 
+ *             the function returns FPGA_INVALID_PARAM.
+ *             When the parameter build_str* to fpgaGetOPAECBuildString() is nullptr, 
+ *             the function returns FPGA_INVALID_PARAM.
+ */
+TEST(open_c_ase, fpga_version) {
+	EXPECT_EQ(FPGA_INVALID_PARAM, fpgaGetOPAECVersion(nullptr));
+	fpga_version version;
+	EXPECT_EQ(FPGA_OK, fpgaGetOPAECVersion(&version));
+
+	EXPECT_EQ(FPGA_INVALID_PARAM, fpgaGetOPAECVersionString(nullptr, 10));
+	EXPECT_EQ(FPGA_INVALID_PARAM, fpgaGetOPAECBuildString(nullptr, 10));
+}
+
+/**
+ * @test       ase_wsid_1
+ *
+ * @brief      When the parameter n_hash_buckets is zero or greater than 16384, 
+ *             the function returns NULL. 
+ */
+TEST(open_c_ase, ase_wsid_1) {
+	EXPECT_EQ(NULL, wsid_tracker_init(0));
+	EXPECT_EQ(NULL, wsid_tracker_init(16385));
+
+}
+
+/**
+* @test       ase_wsid_2
+*
+* @brief      When memory allocation function ase_malloc failed,
+*             wsid_tracker_init() function should return NULL and 
+*             wsid_add() returns false 
+*
+*/
+using namespace opae::testing;
+TEST(sim_sw_ase, ase_wsid_2) {
+	test_system *system_;
+	system_ = test_system::instance();
+	system_->initialize();
+
+    // Invalidate the memory allocation .
+    system_->invalidate_malloc(0, "wsid_tracker_init");
+    system_->invalidate_malloc(0, "ase_malloc");
+    EXPECT_EQ(wsid_tracker_init(1000), (struct wsid_tracker*)NULL);
+
+	system_->finalize();
+}
+
+/**
+ * @test       ase_umsg_1
+ *
+ * @brief      When ase_capability's umsg feature is zero, 
+ *             fpgaGetNumUmsg function returns FPGA_NOT_SUPPORTED. 
+ *             When ase_capability's umsg feature is nonzero,
+ *             fpgaGetNumUmsg function returns FPGA_OK.
+ */
+TEST(open_c_ase, ase_umsg_1) {
+	ase_capability.umsg_feature = 2;
+	struct _fpga_handle handle_;
+	fpga_handle accel_ = &handle_;
+
+	uint64_t value;
+
+	EXPECT_EQ(FPGA_OK, fpgaGetNumUmsg(accel_, &value));
+
+	ase_capability.umsg_feature = 0;
+	EXPECT_EQ(FPGA_NOT_SUPPORTED, fpgaGetNumUmsg(accel_, &value));
+}
+
+/**
+ * @test       ase_umsg_2
+ *
+ * @brief      When ase_capability's umsg feature is zero, 
+ *             fpgaSetUmsgAttributes function returns FPGA_NOT_SUPPORTED. 
+ *             When ase_capability's umsg feature is nonzero,
```

Comment:
This scenario isn't being tested.  Should it be added?

Reply:
Will remove these unsupported feature tests.